pymc3/distributions/multivariate.py

```diff
@@ -486,21 +486,43 @@ def __init__(self, n, p, *args, **kwargs):
 
         p = p / tt.sum(p, axis=-1, keepdims=True)
 
+        self._n = n
         if len(self.shape) == 2:
             try:
                 assert n.shape == (self.shape[0],)
             except AttributeError:
                 # this occurs when n is a scalar Python int or float
                 n *= tt.ones(self.shape[0])
-
             self.n = tt.shape_padright(n)
             self.p = p if p.ndim == 2 else tt.shape_padleft(p)
         else:
             self.n = tt.as_tensor_variable(n)
             self.p = tt.as_tensor_variable(p)
 
         self.mean = self.n * self.p
-        self.mode = tt.cast(tround(self.mean), 'int32')
+        self._raw_mode = tt.cast(tround(self.mean), 'int32')
```

Comment:
Something like this seems to work - could you try to see if it pass the tests?
```python
        rawmode = tt.cast(tround(self.mean), 'int32')
        diff = self.n - tt.sum(rawmode, axis=-1, keepdims=True)
        diffindx = diff.ravel().nonzero()
        newval = rawmode[diffindx, 0] + diff[diffindx].ravel()
        self.mode = tt.set_subtensor(rawmode[diffindx, 0], newval)
```

Reply:
@junpenglao , it seems to work for some shapes but not all.  I think the `ravel` might not be the right thing to do, but I'm still playing with it.

Reply:
I think an argsort needs to occur: you have to figure out which values need 1 added to them and that depends on which are furthest away from the mean.  Even if you want to just do the first n[i] for each row[i], the number could vary per row, there needs to be a way to generate rows with varying numbers of 1s.  I don't know `tensorA[tensorB, 0]` will do that.

Reply:
I think something like this will work if the only criteria is that the sum is preserved
```
mean = n * p
rawmode = tt.cast(tt.round(mean), 'int32')
diff = n - tt.sum(rawmode, axis=-1, keepdims=True)
(n_rows, n_cols) = shape[-2:]
lst = list(range(n_cols) for _ in range(n_rows))
bool_arr = tt.as_tensor_variable(lst) < diff
mode = tt.inc_subtensor(rawmode[bool_arr.nonzero()], 1)
```

I'll keep working on how to find the "proper" indices to increment

Reply:
> you have to figure out which values need 1 added to them and that depends on which are furthest away from the mean.

Yeah I thought about that as well, there will be a bias anyway - especially in a case like `p=[.2, .2]` and `n=1`. So as long as the sum is preserved it is good enough.